README.md
# mapping_explorer
## INSTALLATION
```sh
$ pip install -r requirements.txt
```

```sh
rosrun capture_depth.py
python calib_extrinsic.py

rosrun check_depth.py
```

This is a ROS package. Compile it with catkin.
```sh
mkdir build
cd build
cmake ..
make

```
## Template for commit mdg
What/How  
[New]
[Fix]
[Modi]

## file_list
```
.
├── demo.md  
├── demo.sh
├── bringup_mower.sh  # for set_datum
├── jupyterNotebook
│   ├── Automatic_tree_detection.ipynb
│   ├── icp_svd.ipynb # center of trunk match iteration (quite same as Automatic_xxx)
│   ├── image_util.ipynb
│   ├── spatio_transform_util.ipynb # point cloud distance test
│   ├── ground_subtr_npy.ipynb # using height/depth mask and calcu moment of obj and dist
│   ├── low_pass.ipynb  # moving average to extract ground
│   ├── transformation_height.ipynb  # equal height line and pitch offset for height
│   ├── transformation_pose.ipynb # icp for trunk skin
│   └── histogram_of_trunk.ipynb  # find grass and trunk histogram
│
├── launch
│   ├── rosbagPlayer_w_tf_static_template.launch  # play rosbag with complementary tf tree
│   └── tf_reframe.launch
├── library
│   ├── loadData_utils.py
│   ├── map_utils.py
│   ├── plot_utils.py
│   └── rosbag_utils.py
├── release-notes.md
├── requirements.txt
├── scripts
│   ├── calib_extrinsic.py
│   ├── navsat_tf.py  # use /navsat/fix as tf 
│   ├── cb_pose.py
│   ├── cb_topic2npy.py
│   ├── check_depth.py
│   ├── find_trunk_center_inaframe.py
│   ├── landmark_matcher.py
│   ├── mapping_landmark.py
│   ├── plot_landmark.py
│   ├── tf_remove_frames.py
│   └── rosmsg_depth_template.py  #publish depth.npy to rosmsg
└── src
    ├── depth_2_1dpcl_node.cpp
    └── test.cpp
```

## REFERENCE
[find closest dist](https://www.pyimagesearch.com/2018/07/23/simple-object-tracking-with-opencv/) are a powerful concept. They allow you to, find the *closest point in the dataset*.



### shape file in gps(lat, lon)
![image](https://github.com/wowyunDBL/mapping_explorer/blob/master/image/RGBD-point_cloud.png)



### teach you how to write argument
https://www.pyimagesearch.com/2021/02/22/opencv-connected-component-labeling-and-analysis/


release-notes.md
# Occupancy Mapping Release Notes

## [v0.1]

2021-09-08

- Different mapping layer result
- 

## [v1.0]
2021-11-11

- Analyse cost
```bash
roslaunch mapping_explorer tf_reframe.launch
rosrun mapping_explorer navsat_tf.py
roslaunch depth_app hector_mower.launch
rosrun hector_mapping cbCost.py
```


## Template for commit mdg
What/How
[New]
[Fix]
[Enhancement]

library/plot_utils.py
import rosbag
import math
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm
from matplotlib.ticker import NullFormatter, FixedLocator
from matplotlib import transforms
from pyproj import Proj
import utm

bag_source_file = "/home/ncslaber/110-1/211002_allLibrary/image_bag/2021-10-02-17-54-09.bag"
posBDs_source_file = "/home/ncslaber/109-2/210816_NTU_half/positiveBDs_2021-08-15-11-23-22.bag"

shp_path = '/home/ncslaber/110-1/211002_allLibrary/2021-10-02-17-54-09/shapefiles/'

topics = ['/navsat/fix']#'/navsat/fix'

try:
    ''' Read exploring bag '''
    print('Start reading: {}'.format(bag_source_file))

    lat = []
    lng = []

    bag = rosbag.Bag(bag_source_file, 'r')
    for topic, msg, t in bag.read_messages(topics=topics):
        
        if topic == '/navsat/fix':
            if not math.isnan(msg.latitude) and not math.isnan(msg.longitude):
                lat.append(msg.latitude)
                lng.append(msg.longitude)
            else:
                print("error lat/lon:", msg.latitude, msg.longitude)
    bag.close()

    lat = np.asarray(lat)
    lng = np.asarray(lng)
    gps_xy = np.vstack(( lng, lat))
    gps_xy = np.transpose(gps_xy)
    # np.save('/home/ncslaber/traj_GPS', gps_xy)

    # plot
    fig, ax = plt.subplots(figsize=(12, 12), dpi=100)
    plt.grid(True)
    base = plt.gca().transData
    rot = transforms.Affine2D().rotate_deg(0)#
    
    _, _, zone, _ = utm.from_latlon(lat[0], lng[0])
    print('lat: ', lat.shape)
    proj = Proj(proj='utm', zone=zone, ellps='WGS84', preserve_units=False)
    ux, uy = proj(lng, lat)

    print(np.asarray(ux).shape)
    colors = cm.rainbow(np.linspace(1, 0, np.asarray(ux[:5000]).shape[0]))
    plt.scatter(ux[:5000], uy[:5000], c=colors, transform = rot + base, label="exploring")
    utm_xy = np.vstack(( ux[:5000], uy[:5000]))
    # np.save(bag_source_file[:-4]+'/traj_GPS_filtered', utm_xy)
    ax.axis('equal')
    plt.title("UTM [m]", fontsize=25)

    plt.yticks(fontsize=15)
    plt.xticks(fontsize=15)

    for i in range(8):
        neg_bd = np.load(shp_path+'neg_'+str(i+1)+'_bd_utm.npy')
        center = np.load(shp_path+'center_'+str(i+1)+'_bd_utm.npy')
        if neg_bd is None or center is None:
            print("neg_bd is empty!!")
        plt.scatter(neg_bd[:,0], neg_bd[:,1], c='b', s=10)
        plt.scatter(center[0], center[1], c='b', marker='X',s=100)

    ax.ticklabel_format(useOffset=False, style='sci')
    plt.show()

    ''' Read pos bds bag 
    print('Start reading: {}'.format(posBDs_source_file))
    lat = []
    lng = []

    bag = rosbag.Bag(posBDs_source_file, 'r')
    for topic, msg, t in bag.read_messages(topics=topics):
        
        if topic == '/navsat/fix':
            if not math.isnan(msg.latitude) and not math.isnan(msg.longitude):
                lat.append(msg.latitude)
                lng.append(msg.longitude)
            else:
                print("error lat/lon:", msg.latitude, msg.longitude)
    bag.close()
    print(lat[0],lng[0])
    lat = np.asarray(lat)
    lng = np.asarray(lng)
    gps_xy = np.vstack(( lat, lng))
    gps_xy = np.transpose(gps_xy)

    # plot
    plt.grid(True)
    base = plt.gca().transData
    rot = transforms.Affine2D().rotate_deg(-40)#
    
    _, _, zone, _ = utm.from_latlon(lat[0], lng[0])
    proj = Proj(proj='utm', zone=zone, ellps='WGS84', preserve_units=False)
    ux, uy = proj(lng, lat)

    print(np.asarray(ux).shape)
    # colors = cm.rainbow(np.linspace(1, 0, np.asarray(ux).shape[0]))
    plt.scatter(ux, uy, c='k', transform = rot + base, label="pos_bds")
    utm_xy = np.vstack(( ux, uy))
    # utm_xy = np.transpose(utm_xy)
    np.save('/home/ncslaber/pos_traj_GPS', utm_xy)

    ax.axis('equal')
    plt.title("UTM [m]", fontsize=25)

    plt.yticks(fontsize=15)
    plt.xticks(fontsize=15)

    ax.ticklabel_format(useOffset=False, style='sci')
    plt.legend(fontsize=25)
    #plt.savefig(path+"test_timeMarker.png", dpi=100)
    plt.show()
    '''

except:
    print('error with: {}'.format(bag_source_file)) 